File: clients/typescript/adherence-client.ts
/**
 * adherence-ml client for Med-Tracker (or any TypeScript consumer).
 *
 * Minimal, zero-dependency wrapper around the REST API. Mirrors the
 * payload shapes in packages/common/adherence_common/schemas.py.
 *
 * Usage:
 *   const client = new AdherenceClient({
 *     baseUrl: process.env.ADHERENCE_URL!,
 *     apiKey: process.env.ADHERENCE_SERVICE_KEY!,  // role >= service
 *   });
 *   const res = await client.predict({
 *     user_id: "u_42",
 *     schedule: [
 *       { dose_id: "d1", scheduled_at: "2026-06-01T08:00:00Z",
 *         dose_class: "cardio", dose_strength_mg: 10 },
 *     ],
 *     top_k_reasons: 3,
 *   });
 *
 * For a fully-typed client run:
 *   uv run python scripts/export_openapi.py > openapi.json
 *   npx openapi-typescript openapi.json -o adherence.types.ts
 */

export type DoseClass =
  | "cardio"
  | "psych"
  | "endocrine"
  | "analgesic"
  | "antibiotic"
  | "supplement"
  | "other";

export type RiskTier = "low" | "medium" | "high";

export interface ScheduledDose {
  dose_id: string;
  scheduled_at: string; // ISO-8601 UTC
  dose_class: DoseClass;
  dose_strength_mg: number;
}

export interface DoseHistoryEvent {
  user_id: string;
  dose_id: string;
  scheduled_at: string;
  taken_at?: string | null;
  status: "taken" | "missed" | "late" | "skipped";
  dose_class: DoseClass;
  dose_strength_mg: number;
}

export interface PredictRequest {
  user_id: string;
  schedule: ScheduledDose[];
  history?: DoseHistoryEvent[];
  top_k_reasons?: number;
}

export interface ReasonCode {
  feature: string;
  contribution: number;
  human: string;
}

export interface DosePrediction {
  dose_id: string;
  scheduled_at: string;
  miss_probability: number;
  risk_tier: RiskTier;
  reasons: ReasonCode[];
}

export interface PredictResponse {
  user_id: string;
  model_version: string;
  predictions: DosePrediction[];
}

export interface AdherenceClientOptions {
  baseUrl: string;
  apiKey?: string;
  bearerToken?: string;
  timeoutMs?: number;
  fetchImpl?: typeof fetch;
}

export class AdherenceClient {
  private baseUrl: string;
  private apiKey?: string;
  private bearerToken?: string;
  private timeoutMs: number;
  private fetchImpl: typeof fetch;

  constructor(opts: AdherenceClientOptions) {
    this.baseUrl = opts.baseUrl.replace(/\/$/, "");
    this.apiKey = opts.apiKey;
    this.bearerToken = opts.bearerToken;
    this.timeoutMs = opts.timeoutMs ?? 5000;
    this.fetchImpl = opts.fetchImpl ?? fetch;
  }

  private headers(): Record<string, string> {
    const h: Record<string, string> = { "content-type": "application/json" };
    if (this.apiKey) h["x-api-key"] = this.apiKey;
    if (this.bearerToken) h["authorization"] = `Bearer ${this.bearerToken}`;
    return h;
  }

  private async req<T>(method: string, path: string, body?: unknown): Promise<T> {
    const ctrl = new AbortController();
    const t = setTimeout(() => ctrl.abort(), this.timeoutMs);
    try {
      const r = await this.fetchImpl(`${this.baseUrl}${path}`, {
        method,
        headers: this.headers(),
        body: body === undefined ? undefined : JSON.stringify(body),
        signal: ctrl.signal,
      });
      if (!r.ok) {
        const text = await r.text().catch(() => "");
        throw new Error(`adherence-ml ${method} ${path} -> ${r.status}: ${text}`);
      }
      return (await r.json()) as T;
    } finally {
      clearTimeout(t);
    }
  }

  health(): Promise<{ status: string; model_loaded: boolean; version: string }> {
    return this.req("GET", "/healthz");
  }

  predict(req: PredictRequest, modelName = "default"): Promise<PredictResponse> {
    const q = `?model_name=${encodeURIComponent(modelName)}`;
    return this.req("POST", `/v1/predict${q}`, req);
  }

  cohortRisk(
    payload: { events?: DoseHistoryEvent[]; synthetic?: { n_users?: number; n_days?: number; seed?: number } },
    opts: { modelName?: string; topUsers?: number } = {},
  ): Promise<unknown> {
    const params = new URLSearchParams();
    params.set("model_name", opts.modelName ?? "default");
    if (opts.topUsers) params.set("top_users", String(opts.topUsers));
    return this.req("POST", `/v1/cohort/risk?${params.toString()}`, payload);
  }

  explainGlobal(modelName = "default"): Promise<unknown> {
    return this.req("GET", `/v1/explain/global?model_name=${encodeURIComponent(modelName)}`);
  }
}

File: clients/python/adherence_client.py
"""Minimal Python client (sync, stdlib + httpx) for adherence-ml.

Mirrors the TypeScript client in ../typescript/adherence-client.ts so that
Med-Tracker server-side jobs (Python) and the Med-Tracker web app
(TypeScript) talk to the same surface.
"""
from __future__ import annotations

from dataclasses import dataclass
from typing import Any, Iterable

import httpx


@dataclass
class AdherenceClient:
    base_url: str
    api_key: str | None = None
    bearer_token: str | None = None
    timeout_s: float = 5.0

    def _headers(self) -> dict[str, str]:
        h = {"content-type": "application/json"}
        if self.api_key:
            h["x-api-key"] = self.api_key
        if self.bearer_token:
            h["authorization"] = f"Bearer {self.bearer_token}"
        return h

    def _req(self, method: str, path: str, **kw: Any) -> Any:
        url = self.base_url.rstrip("/") + path
        with httpx.Client(timeout=self.timeout_s, headers=self._headers()) as c:
            r = c.request(method, url, **kw)
            r.raise_for_status()
            return r.json()

    def health(self) -> dict[str, Any]:
        return self._req("GET", "/healthz")

    def predict(
        self,
        user_id: str,
        schedule: Iterable[dict[str, Any]],
        history: Iterable[dict[str, Any]] | None = None,
        model_name: str = "default",
        top_k_reasons: int = 3,
    ) -> dict[str, Any]:
        body = {
            "user_id": user_id,
            "schedule": list(schedule),
            "history": list(history) if history is not None else None,
            "top_k_reasons": top_k_reasons,
        }
        return self._req(
            "POST", f"/v1/predict?model_name={model_name}", json=body
        )

    def predict_batch(
        self,
        items: Iterable[dict[str, Any]],
        model_name: str = "default",
    ) -> dict[str, Any]:
        return self._req(
            "POST",
            f"/v1/predict/batch?model_name={model_name}",
            json={"items": list(items)},
        )

    def cohort_risk(
        self,
        events: Iterable[dict[str, Any]] | None = None,
        synthetic: dict[str, Any] | None = None,
        model_name: str = "default",
        top_users: int = 10,
    ) -> dict[str, Any]:
        body: dict[str, Any] = {}
        if events is not None:
            body["events"] = list(events)
        if synthetic is not None:
            body["synthetic"] = synthetic
        return self._req(
            "POST",
            f"/v1/cohort/risk?model_name={model_name}&top_users={top_users}",
            json=body,
        )

    def explain_global(self, model_name: str = "default") -> dict[str, Any]:
        return self._req("GET", f"/v1/explain/global?model_name={model_name}")
